### PR TITLE
AI shooting bots

### PR DESCRIPTION
**Pull request by ghost**
Bots can:
* reload the weapon when needed
* shoot the player
* run away from the player

Bots can't:
* avoid obstacles
* hide from player

#24

### CODE REVIEW

File: Source/Prot/AI/FireBTTaskNode.cpp [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+
+EBTNodeResult::Type UFireBTTaskNode::ExecuteTask(UBehaviorTreeComponent & OwnerComp, uint8 * NodeMemory)
+{
+	UBlackboardComponent* BlackboardComponent = OwnerComp.GetBlackboardComponent();
+	if (BlackboardComponent)
+	{
+		ABot* Bot = Cast<ABot>(BlackboardComponent->GetValueAsObject("SelfActor"));
+		if (Bot)
+		{
+			this->Bot = Bot;
```
Comment (UnoYakshi) [inline Source/Prot/AI/FireBTTaskNode.cpp]: The hell is that?..

File: Source/Prot/Prot.h [hunk 2]
```diff
@@ -13,4 +13,4 @@
 #define COLLISION_PICKUP		ECC_GameTraceChannel3
 
 // Either print debug-lines on screen or not...
-#define DEBUG false
+#define DEBUG true
```
Comment (UnoYakshi) [inline Source/Prot/Prot.h]: Switch it back.

File: Source/Prot/Bot.h [hunk 3]
```diff
@@ -25,15 +26,17 @@ class PROT_API ABot : public ACharacter
...
 	// Points that Bot will patrol in the same order as they are in the array
 	UPROPERTY(EditAnywhere, BlueprintReadWrite, Category = "AI")
 	TArray<ATargetPoint*> PatrolPoints;
 
-	//Is patrolling order random?
+	// Is patrolling order random?
 	UPROPERTY(EditAnywhere, BlueprintReadWrite, Category = "AI")
 	bool bPatrolRandomly;
+
+	AWeapon* GetCurrentWeapon() { return CurrentWeapon; }
```
Comment (UnoYakshi) [inline Source/Prot/Bot.h]: We have `AWeapon* CurrentWeapon` [to be `BlueprintReadWrite`] in `ProtCharacter`. Move the getter there. Also add a specifier to `CurrentWeapon`: `BlueprintGetter = GetCurrentWeapon`.

File: Source/Prot/Bot.h [hunk 4]
```diff
@@ -3,6 +3,7 @@
 #pragma once
 
 #include "CoreMinimal.h"
+#include "ProtCharacter.h"
 #include "GameFramework/Character.h"
```
Comment (UnoYakshi) [inline Source/Prot/Bot.h]: Remove `Character.h` as we don't need it anymore, for now.

File: Source/Prot/AI/ReloadBTTaskNode.cpp [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+
+EBTNodeResult::Type UReloadBTTaskNode::ExecuteTask(UBehaviorTreeComponent & OwnerComp, uint8 * NodeMemory)
+{
+	UBlackboardComponent* BlackboardComponent = OwnerComp.GetBlackboardComponent();
+	if (BlackboardComponent)
+	{
+		ABot* Bot = Cast<ABot>(BlackboardComponent->GetValueAsObject("SelfActor"));
+		if (Bot)
+		{
+			if (Bot->GetCurrentWeapon())
```
Comment (UnoYakshi) [inline Source/Prot/AI/ReloadBTTaskNode.cpp]: Probably worth do move it to the first condition so it takes less visual space.

File: Source/Prot/AI/FireBTTaskNode.h [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+ */
+UCLASS()
+class PROT_API UFireBTTaskNode : public UBTTaskNode
+{
+	GENERATED_BODY()
+	// Handle to manage bot shooting timer
+	FTimerHandle FireTimerHandle;
+	ABot* Bot;
+	virtual void StopShooting();
+	virtual EBTNodeResult::Type ExecuteTask(UBehaviorTreeComponent& OwnerComp, uint8* NodeMemory) override;
```
Comment (UnoYakshi) [inline Source/Prot/AI/FireBTTaskNode.h:29]: Add newlines, and comments to all of those.

File: Source/Prot/AI/BotAIC.cpp [hunk 7]
```diff
@@ -29,9 +30,19 @@ void ABotAIC::Possess(APawn* Pawn)
 			BlackboardComponent->InitializeBlackboard(*(Bot->BehaviorTree->BlackboardAsset));
 			BlackboardComponent->SetValueAsInt("PointIndex", 0);
 
+			// Set Bot that uses the blackboard
+			BlackboardComponent->SetValueAsObject("SelfActor", Bot);
+
 			// Start the behavior tree which corresponds to the specific character
 			BehaviorTreeComponent->StartTree(*Bot->BehaviorTree);
 
+			UHealthActorComponent* BotHealhComponent =
```
Comment (UnoYakshi) [inline Source/Prot/AI/BotAIC.cpp:40]: Probably, worth to comment on this area.

**Review by UnoYakshi — CHANGES_REQUESTED** [review]

File: Source/Prot/AI/FireBTTaskNode.cpp (hunk 1 above)
Comment (ghost) [reply Source/Prot/AI/FireBTTaskNode.cpp]: What's wrong?

Comment (UnoYakshi) [reply Source/Prot/AI/FireBTTaskNode.cpp]: Namings. It's not a constructor to perform such things.

File: Source/Prot/AI/BotAIC.cpp (hunk 7 above)
Comment (andrew-it) [reply Source/Prot/AI/BotAIC.cpp:40]: @UnoYakshi, why?

**Review by andrew-it — APPROVED** [review]

File: Source/Prot/AI/BotAIC.cpp (hunk 7 above)
Comment (UnoYakshi) [reply Source/Prot/AI/BotAIC.cpp:40]: So you won't read all the `AddDynamic`, `::`, etc.

File: Source/Prot/AI/FireBTTaskNode.h (hunk 6 above)
Comment (UnoYakshi) [reply Source/Prot/AI/FireBTTaskNode.h:29]: Add some newlines, will you?

File: Source/Prot/ProtCharacter.h [hunk 8]
```diff
@@ -265,5 +265,6 @@ class AProtCharacter : public ACharacter
 	/** Returns Weapon attach point... */
 	FName GetWeaponAttachPoint() const;
 
+	// Gets current weapon
+	AWeapon* GetCurrentWeapon() { return CurrentWeapon; }
```
Comment (UnoYakshi) [inline Source/Prot/ProtCharacter.h:270]: Have you tested it inside the Editor w/o `UPROPERTY(BlueprintGetter)`?

**Review by UnoYakshi — CHANGES_REQUESTED** [review]
If the future, consider to write documentation for tasks' classes. They are kind of obvious for now, but they are simplistic. We don't know if they will grow up or not.

**Review by UnoYakshi — APPROVED** [review]

**Pull request merged**